tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15358,10 +15358,66 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
   return;
 }
 
+// if this is the Entry FD, then try adding the target profile
+// shader attribute to the FD and carry on with validation
+void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD) {
+  const std::string &EntryPointName = S.getLangOpts().HLSLEntryFunction;  
+  
+  // if there's no defined entry point, just return
+  if (EntryPointName.empty()) {
+    return;
+  }
+  
+  // There's no way we should be adding a shader attribute
+  // to a Function Decl if it doesn't even have an identifier
+  if (!FD->getIdentifier()) {
+    return;
+  }
+
+  // if this FD isn't the entry point, then we shouldn't add
+  // a shader attribute to this decl, so just return
+  if (EntryPointName != FD->getIdentifier()->getName()) {
+    return;
+  }
+
+  std::string profile = S.getLangOpts().HLSLProfile;
+  const ShaderModel *SM = hlsl::ShaderModel::GetByName(profile.c_str());
+  const llvm::StringRef fullName = ShaderModel::FullNameFromKind(SM->GetKind());
+
+  // don't add the attribute for an invalid profile, like library
+  if (fullName.empty()) {
+    return;
+  }
+
+  HLSLShaderAttr *currentShaderAttr = FD->getAttr<HLSLShaderAttr>();
+  // don't add the attribute if it already exists as an attribute on the decl  
+  if (currentShaderAttr) {
+    llvm::StringRef currentFullName = currentShaderAttr->getStage();
+    if (currentFullName == fullName) {
+      return;
+    }
+  }
+
+  // in the special case that the target profile is compute and the 
+  // entry decl already has a node shader attr, don't do anything
+  if (fullName == "compute" && currentShaderAttr->getStage() == "node") {
```

Comment:
I think the complication is necessary when considering the responsibility of the function ("Try to add the target profile as a shader attribute to the entry function") We shouldn't do this if the target is compute and there's already a node attribute, that check seems necessary to place in this function to me. 

I do see the problem that there could already be several attributes on the decl, and that we can choose the wrong one. However, the multiple attributes currently will be caught by the code in CGHLSLMS.cpp, and those diags will eventually be moved to Sema, so I don't think this function has anything wrong specifically in terms of design. To be clear, I don't think this function's responsibilties should involve checking to see that the entry decl has exactly one attribute. I've updated the code to prevent conflicting attributes from being added, along with a note that tells the user that the target profile conflicts with an existing explicit attribute. As to the multiple attributes problem, I think this function is fine and I should make an issue to move the multiple attributes detection in CGHLSLMS to a location before this function gets called.